README
BUILDING, TESTING, INSTALLING
=============================

Unix
----

Using Autotools:
  ./autogen.sh  # in ${srcdir}, if obtained from SVN
  ${srcdir}/configure # in build dir

Using CMake:
  cmake ${srcdir} # in build dir

Now, all versions:
  make
  make check
  make install # as root
  ldconfig # as root

Win32
-----

If you use Microsoft Visual C++ (7.1 or later) compiler, you can build
GEOS using NMAKE program and provided makefile.vc files.

If you are building from SVN checkout, first run: autogen.bat
Then:

  nmake /f makefile.vc MSVC_VER=1400

where 1400 is version number of Visual C++ compiler, here Visual C++ 8.0
from Visual Studio 2005 (supported versions are 1300, 1310, 1400 and 1500).
The bootstrap.bat step is required to generate a couple of header files.

In order to build debug configuration of GEOS, additional flag DEBUG=1
is required:

  nmake /f makefile.vc MSVC_VER=1400 DEBUG=1


CLIENT APPLICATIONS
===================

Using the C interface (recommended)
-----------------------------------

To compile programs against the C lib (recommended):

  CFLAGS += `geos-config --cflags`
  LDFLAGS += `geos-config --ldflags` -lgeos_c
  #include <geos_c.h>

Example usage:

  capi/geostest.c contains basic usage examples.

Using the C++ interface (discouraged)
-------------------------------------

To compile programs against the C++ lib:

  CFLAGS += `geos-config --cflags`
  LDFLAGS += `geos-config --ldflags` -lgeos
  #include <geos.h>

Example usage:

  doc/example.cpp contains basic usage examples.

Scripting language bindings
---------------------------

Ruby bindings are fully supported. To build, use the --enable-ruby option
when configuring:

  ./configure ... --enable-ruby

PHP bindings are fully supported. To build, use the --enable-php option
when configuring:

  ./configure ... --enable-php

Since version 3.0, the Python bindings are unsupported. Recommended options:

 1. Become or recruit a new maintainer.
 2. Use Shapely (http://pypi.python.org/pypi/Shapely) with Python
    versions 2.4 or greater.
 3. Simply call functions from libgeos_c via Python ctypes.

DOCUMENTATION
=============

  cd doc; make doxygen-html

  http://trac.osgeo.org/geos



NEWS
Changes in 3.4.0
????-??-??

- C++ API changes:
  - Added BufferOp::setSingleSided
  - Signature of most functions taking a Label changed to take it
    by reference rather than pointer.
  - Signature of most functions taking an IntersectionMatrix changed
    to take it by reference rather than pointer.
  - GraphComponent::label is now a Label value (from a pointer)
  - NodedSegmentString takes ownership of CoordinateSequence now
- Bug fixes / improvements
  - Fixed Linear Referencing API to handle MultiLineStrings consistently
    by always using the lowest possible index value, and by trimming
    zero-length components from results (#323)

Changes in 3.3.0
2011-05-30

- New things:
  - CAPI: GEOSBufferWithParams (allows single sided buffers)
  - CAPI: GEOSOffsetCurve deprecates GEOSSingleSidedBuffer
  - CAPI: GEOSUnaryUnion deprecates GEOSCascadedUnion
  - CAPI: GEOSisValidDetail: tell state, reason & location apart. allows
          passing flags.
  - CAPI: GEOSContext_setNoticeHandler_r, GEOSContext_setErrorHandler_r
  - CAPI: GEOSGeom_createEmptyPoint, GEOSGeom_createEmptyLineString
          GEOSGeom_createEmptyPolygon, GEOSGeom_createEmptyCollection
  - CAPI: GEOSGeom_extractUniquePoints
  - CAPI: GEOSGetGeometryN support for single geometries
  - CAPI: GEOSPolygonize_full to return all informations computed by
          the polygonizer
  - CAPI: GEOSOrientationIndex
  - CAPI: GEOSSharedPaths to find shared paths and their orientation
  - CAPI: GEOSSnap
  - CAPI: GEOSRelatePatternMatch
  - CAPI: GEOSCovers, GEOSCoveredBy (#396)
  - CAPI: GEOSRelateBoundaryNodeRule (#399)
  - CAPI: GEOSGeom_getCoordinateDimension() (#311)
  - CAPI: GEOSWKTWriter_setOutputDimension, GEOSWKTWriter_setOld3D (#292)
  -  PHP: new PHP5 bindings based on CAPI
- Semantic C++ API changes:
  - Geometry inheritance chain changed to introduce Puntal, Lineal
    and Polygonal classes (virtual inheritance introduced)
  - Polygonizer::getInvalidRingLines retains ownership of vector elements
  - Geometry::isWithinDistance method is now const
  - Polygonizer::getCutEdges returns by const ref
  - Polygonizer::getDangles returns by const ref
  - Empty LinearRings are closed by definition
- Bug fixes / improvements
  - Fixed Geometry.distance() and DistanceOp to return 0.0 for empty inputs
  - Invalid compound geometries reported as valid (#333)
  - Return up to 15 digits of precision from GEOSisValidReason_t (#329)
  - CAPI: do not leak contexts when using the non-reentrant interface
  - Fix duplicated dangles returned by Polygonizer
  - Fix SnapIfNeededOverlayOp to throw the originating exception
  - Fixed LineMerger to skip lines with only a single unique coordinate
  - Fix NodedSegmentString to handle zero-length line segments correctly
    (via safeOctant)
  - Fixed buffer OffsetCurveSetBuilder to handle "flat" rings correctly
  - Added illegal state check in LineSegment::pointAlongOffset()
  - Improved performance of RectangleIntersects by always using
    segment-scanning and refining SegmentIntersectionTester
  - Reduce memory use in PreparedGeometry predicates (#342)
  - Fix infinite loop in RobustDeterminant with nan/inf input (#357)

Changes in 3.2.0
2009-12-14

- Add Single-sided buffer operation
- JTS-1.10 sync ...
  - Drastically improved Buffer speed (20x for a case in testsuite)
  - Improved EdgeList duplicate edge finding
  - Added algorithm::distance package
  - Added algorithm::Angle class
  - Added algorithm::BoundaryNodeRule class
  - IsSimpleOp can now return non-simple location coordinate
  - DistanceOp can now check for 'within distance' predicate
    (earlier exit)
  - MultiPolygon::getBoundary always return MultiLineString, also
    when the result is the EMPTY geometry.
- Various bug and leak fixes, optimizations
- Replace MarkupSTL with tinyXML

Changes in 3.1.0

- PreparedGeometry operations for very fast predicate testing.
  - Intersects()
  - Covers()
  - CoveredBy()
  - ContainsProperly()
- Easier builds under MSVC and OpenSolaris
- Thread-safe CAPI option
- IsValidReason added to CAPI
- GEOSPolygonizer_getCutEdges added to CAPI
- CascadedUnion operation for fast unions of geometry sets
- Numerous bug fixes.
  http://trac.osgeo.org/geos/query?status=closed&milestone=3.1.0&order=priority

Changes in 3.0.0

   These are mostly ABI breaking changes.
   In few cases the API also changed, but the most external one
   (the documented one) should be unchanged.

- New things:
	- Added geom::BinaryOp class performing a binary operation
	  using different heuristics to reduce probability of robustness
	  issues. Both C-API and XMLTester now use this class for
	  binary operations.
	- Added covers() and coveredBy() predicates to Geometry class
	- Added overlay::overlayOp() adapter class
	- Added GEOSSimplify() and GEOSTopologyPreserveSimplify()
	  to the C API
	- Added closed ring checks in IsValidOp
	- Multi-input support in XMLTester
	- HEXWKB I/O
	- Envelope(string) ctor
	- Ruby interface
	- New ShortCircuitedGeometryVisitor class
	- New operation/predicate package
	- Added CGAlgorithms::isPointInRing() version working with
	  Coordinate::ConstVect type (faster!)
	- Added getAt(int pos, Coordinate &to) funtion to CoordinateSequence
	  class.
	- Moved GetNumGeometries() and GetGeometryN() interfaces
	  from GeometryCollection to Geometry class.
	- New planarSubgraph class
	- New ConnectedSubgraphFinder class.
	- New LineSequencer class
	- New WKTWriter::toLineString and ::toPoint convenience methods
	- New IsValidOp::setSelfTouchingRingFormingHoleValid method
	- New WKTWriter::toLineString and ::toPoint convenience methods
	- New IsValidOp::setSelfTouchingRingFormingHoleValid method
	- New Envelope::centre()
	- New Envelope::intersection(Envelope)
	- New Envelope::expandBy(distance, [ydistance])
	- New LineString::reverse()
	- New MultiLineString::reverse()
	- New Geometry::buffer(distance, quadSeg, endCapStyle)
	- New SnapRounding code

src/io/Unload.cpp
/**********************************************************************
 *
 * GEOS - Geometry Engine Open Source
 * http://geos.osgeo.org
 *
 * Copyright (C) 2005-2006 Refractions Research Inc.
 * Copyright (C) 2001-2002 Vivid Solutions Inc.
 *
 * This is free software; you can redistribute and/or modify it under
 * the terms of the GNU Lesser General Public Licence as published
 * by the Free Software Foundation.
 * See the COPYING file for more information.
 *
 **********************************************************************/

#include <geos/unload.h>
#include <geos/geom/Geometry.h>
#include <geos/geom/GeometryFactory.h>

namespace geos {
namespace io { // geos.io

/*public static*/
void Unload::Release()
{
	//delete geom::Geometry::INTERNAL_GEOMETRY_FACTORY;
}

} // namespace geos.io
} //namespace geos

/**********************************************************************
 * $Log$
 * Revision 1.26  2006/04/10 13:09:49  strk
 * Added GeometryFactory::defaultInstance()
 * Made Geometry::INTERNAL_GEOMETRY_FACTORY an alias for it
 * removed last deletion from Unload::Release class
 *
 * Revision 1.25  2006/03/20 16:57:44  strk
 * spatialindex.h and opValid.h headers split
 *
 * Revision 1.24  2006/03/17 16:48:55  strk
 * LineIntersector and PointLocator made complete components of RelateComputer
 * (were statics const pointers before). Reduced inclusions from opRelate.h
 * and opValid.h, updated .cpp files to allow build.
 *
 * Revision 1.23  2006/03/06 19:40:47  strk
 * geos::util namespace. New GeometryCollection::iterator interface, many cleanups.
 *
 * Revision 1.22  2006/03/06 15:23:14  strk
 * geos::io namespace
 *
 * Revision 1.21  2006/03/03 10:46:21  strk
 * Removed 'using namespace' from headers, added missing headers in .cpp files, removed useless includes in headers (bug#46)
 *
 * Revision 1.20  2006/03/02 14:34:43  strk
 * GeometryGraphOperation::li made a non-static member, and not more a pointer
 *
 * Revision 1.19  2006/03/01 10:39:59  strk
 * ctor and dtor made private and inlined
 *
 * Revision 1.18  2006/02/23 11:54:20  strk
 * - MCIndexPointSnapper
 * - MCIndexSnapRounder
 * - SnapRounding BufferOp
 * - ScaledNoder
 * - GEOSException hierarchy cleanups
 * - SpatialIndex memory-friendly query interface
 * - GeometryGraph::getBoundaryNodes memory-friendly
 * - NodeMap::getBoundaryNodes memory-friendly
 * - Cleanups in geomgraph::Edge
 * - Added an XML test for snaprounding buffer (shows leaks, working on it)
 *
 * Revision 1.17  2006/02/20 10:14:18  strk
 * - namespaces geos::index::*
 * - Doxygen documentation cleanup
 *
 * Revision 1.16  2006/02/19 19:46:49  strk
 * Packages <-> namespaces mapping for most GEOS internal code (uncomplete, but working). Dir-level libs for index/ subdirs.
 *
 * Revision 1.15  2006/01/31 19:07:34  strk
 * - Renamed DefaultCoordinateSequence to CoordinateArraySequence.
 * - Moved GetNumGeometries() and GetGeometryN() interfaces
 *   from GeometryCollection to Geometry class.
 * - Added getAt(int pos, Coordinate &to) funtion to CoordinateSequence class.
 * - Reworked automake scripts to produce a static lib for each subdir and
 *   then link all subsystem's libs togheter
 * - Moved C-API in it's own top-level dir capi/
 * - Moved source/bigtest and source/test to tests/bigtest and test/xmltester
 * - Fixed PointLocator handling of LinearRings
 * - Changed CoordinateArrayFilter to reduce memory copies
 * - Changed UniqueCoordinateArrayFilter to reduce memory copies
 * - Added CGAlgorithms::isPointInRing() version working with
 *   Coordinate::ConstVect type (faster!)
 * - Ported JTS-1.7 version of ConvexHull with big attention to
 *   memory usage optimizations.
 * - Improved XMLTester output and user interface
 * - geos::geom::util namespace used for geom/util stuff
 * - Improved memory use in geos::geom::util::PolygonExtractor
 * - New ShortCircuitedGeometryVisitor class
 * - New operation/predicate package
 *
 * Revision 1.14  2004/12/08 14:30:53  strk
 * Added default profiler instance cleanup
 *
 * Revision 1.13  2004/10/13 10:03:02  strk
 * Added missing linemerge and polygonize operation.
 * Bug fixes and leaks removal from the newly added modules and
 * planargraph (used by them).
 * Some comments and indentation changes.
 *
 * Revision 1.12  2004/07/08 19:34:49  strk
 * Mirrored JTS interface of CoordinateSequence, factory and
 * default implementations.
 * Added CoordinateArraySequenceFactory::instance() function.
 *
 * Revision 1.11  2004/07/02 13:28:27  strk
 * Fixed all #include lines to reflect headers layout change.
 * Added client application build tips in README.
 *
 * Revision 1.10  2004/05/05 10:54:48  strk
 * Removed some private static heap explicit allocation, less cleanup done by
 * the unloader.
 *
 * Revision 1.9  2004/04/16 08:52:52  strk
 * Unload::Release final delete (static heap allocations should be gone now)
 *
 * Revision 1.8  2004/04/16 07:42:06  strk
 * PrecisionModel::Type made an enum instead of a Type.
 *
 * Revision 1.7  2004/04/15 15:11:38  strk
 * Commented out deletion that seems to cause segfaults
 *
 * Revision 1.6  2004/04/15 14:00:31  strk
 * Added new cleanup to Unload::Release
 *
 * Revision 1.5  2004/04/01 10:44:34  ybychkov
 * All "geom" classes from JTS 1.3 upgraded to JTS 1.4
 *
 * Revision 1.4  2004/03/29 06:59:25  ybychkov
 * "noding/snapround" package ported (JTS 1.4);
 * "operation", "operation/valid", "operation/relate" and "operation/overlay" upgraded to JTS 1.4;
 * "geom" partially upgraded.
 *
 * Revision 1.3  2004/03/19 09:48:46  ybychkov
 * "geomgraph" and "geomgraph/indexl" upgraded to JTS 1.4
 *
 * Revision 1.2  2003/11/07 01:23:42  pramsey
 * Add standard CVS headers licence notices and copyrights to all cpp and h
 * files.
 *
 *
 **********************************************************************/

